File: web/lib/types/workout-stats.ts
// Wire types for the web Overview read-model (IMPL-WEB-WORKOUT-01 §5). Mirrors
// the backend com.gte619n.healthfitness.core.workoutstats records + the
// SessionDetailResponse. Import-safe from client and server.

import type { ScheduledWorkoutResponse } from "./workout-program";

export type WorkoutStreak = {
  current: number;
  longest: number;
  weeklyTarget: number;
  thisWeekCompleted: number;
  weekStart: string; // ISO date (Monday)
};

export type WeekPoint = {
  weekStart: string; // ISO date (Monday)
  sessions: number;
  tonnageLbs: number;
};

export type SessionRef = {
  programId: string;
  scheduledId: string;
};

export type HeatmapDay = {
  date: string; // ISO date
  sessionCount: number;
  first: SessionRef;
  // IMPL-DELOAD-01 (D4): the day's representative session was a scheduled deload.
  isDeload: boolean;
};

export type PrPoint = {
  exerciseId: string;
  exerciseName: string;
  e1rmLbs: number;
  weightLbs: number;
  reps: number | null;
  date: string; // ISO date
  programId: string;
  scheduledId: string;
  // IMPL-PROG-LOAD-01 (D3/D8): 1 or 2, and pre-doubled TOTAL values for per-hand
  // lifts (dumbbell / dual-cable). Web renders the totals.
  loadFactor: number;
  e1rmTotalLbs: number;
  weightTotalLbs: number;
};

export type LiftRef = {
  exerciseId: string;
  exerciseName: string;
};

export type TrackedExercise = {
  exerciseId: string;
  exerciseName: string;
  lastPerformed: string | null;
};

export type WorkoutStats = {
  streak: WorkoutStreak;
  weeklySeries: WeekPoint[];
  heatmap: HeatmapDay[];
  recentPrs: PrPoint[];
  chartDefaultLifts: LiftRef[];
  trackedExercises: TrackedExercise[];
};

export type E1rmPoint = {
  date: string; // ISO date
  e1rmLbs: number;
  weightLbs: number | null;
  reps: number | null;
  lowConfidence: boolean;
  // IMPL-PROG-LOAD-01: pre-doubled TOTAL values for per-hand lifts.
  e1rmTotalLbs: number;
  weightTotalLbs: number | null;
};

export type E1rmBelief = {
  e1rmLbs: number;
  sigmaLbs: number;
  confidence: string; // HIGH | MEDIUM | LOW
};

export type E1rmHistory = {
  exerciseId: string;
  exerciseName: string;
  points: E1rmPoint[];
  currentBelief: E1rmBelief | null;
  // IMPL-PROG-LOAD-01: 1 for total-load lifts, 2 for per-hand. Multiply the
  // belief by this; the points already carry pre-doubled totals.
  loadFactor: number;
};

// A pointer to an adjacent session for prev/next navigation on the detail page.
export type NeighborRef = {
  programId: string;
  scheduledId: string;
  date: string;
};

// GET /api/me/workout-programs/{programId}/sessions/{scheduledId} — the full
// performed session plus the PR set keys to badge and the neighbors to page to.
export type SessionDetailResponse = {
  session: ScheduledWorkoutResponse;
  prSetKeys: string[]; // "blockId:orderIndex:setIndex"
  prev: NeighborRef | null;
  next: NeighborRef | null;
};

File: backend/src/main/java/com/gte619n/healthfitness/core/workoutstats/WorkoutStats.java
package com.gte619n.healthfitness.core.workoutstats;

import java.time.LocalDate;
import java.util.List;

/**
 * The read-model powering the web Overview dashboard (IMPL-WEB-WORKOUT-01):
 * consistency streak, weekly tonnage/session series, a day-level heatmap,
 * recent personal records, and the exercise lists the strength chart needs.
 * All of it is derived per-request from the user's COMPLETED
 * {@link com.gte619n.healthfitness.core.workoutprogram.ScheduledWorkout}s (the
 * expensive Firestore scan is cached; the derivation is a cheap fold).
 *
 * @param streak            consecutive-weeks streak + current-week progress
 * @param weeklySeries      exactly {@code weeks} entries, oldest→newest, the
 *                          current week last, zero-filled for weeks with no work
 * @param heatmap           only days with ≥1 completed session, last ~6 months
 * @param recentPrs         newest-first, capped at 5 (see {@link WorkoutStatsService})
 * @param chartDefaultLifts the lifts the strength chart defaults to (top by
 *                          observation count, ≤1 per movement pattern)
 * @param trackedExercises  every exercise with a plottable trend — logged on ≥2
 *                          sessions — newest-performed first (picker population)
 */
public record WorkoutStats(
    Streak streak,
    List<WeekPoint> weeklySeries,
    List<HeatmapDay> heatmap,
    List<PrPoint> recentPrs,
    List<LiftRef> chartDefaultLifts,
    List<TrackedExercise> trackedExercises
) {

    /**
     * @param current            consecutive ISO weeks (Monday start) meeting the
     *                           weekly target, including the current in-progress
     *                           week when it already meets the target
     * @param longest            the longest such run over all history
     * @param weeklyTarget       completed workouts/week required (the setting)
     * @param thisWeekCompleted  completed sessions so far in the current week
     * @param weekStart          the current week's Monday (caller's local zone)
     */
    public record Streak(int current, int longest, int weeklyTarget, int thisWeekCompleted, LocalDate weekStart) {}

    /** One week of the volume series. {@code tonnageLbs} is Σ(weight×reps). */
    public record WeekPoint(LocalDate weekStart, int sessions, double tonnageLbs) {}

    /**
     * One workout day for the consistency heatmap. {@code isDeload} is true when
     * the day's representative session was a scheduled deload (IMPL-DELOAD-01 D4).
     */
    public record HeatmapDay(LocalDate date, int sessionCount, SessionRef first, boolean isDeload) {}

    /** A stable pointer to one performed session (for deep-linking). */
    public record SessionRef(String programId, String scheduledId) {}

    /**
     * A personal-record session: a new best estimated 1RM for an exercise.
     *
     * <p>IMPL-PROG-LOAD-01 (D3/D8): {@code e1rmLbs}/{@code weightLbs} are the raw
     * logged (per-hand) numbers; {@code loadFactor} (1 or 2) and the pre-doubled
     * {@code e1rmTotalLbs}/{@code weightTotalLbs} express total load so dumbbell
     * lifts compare to barbell lifts. Web renders the totals.
     */
    public record PrPoint(
        String exerciseId, String exerciseName,
        double e1rmLbs, double weightLbs, Integer reps,
        LocalDate date, String programId, String scheduledId,
        int loadFactor, double e1rmTotalLbs, double weightTotalLbs) {}

    /** An exercise the strength chart can plot, with its display name. */
    public record LiftRef(String exerciseId, String exerciseName) {}

    /** An exercise with a plottable strength trend (≥2 sessions), for the chart's lift picker. */
    public record TrackedExercise(String exerciseId, String exerciseName, LocalDate lastPerformed) {}
}
